File: lib/push.ts
/**
 * Push Notification 공유 유틸리티
 *
 * web-push를 사용하여 사용자에게 Push 알림을 발송합니다.
 * app/api/push/send/route.ts와 app/api/requests/route.ts 양쪽에서 사용합니다.
 */

// eslint-disable-next-line @typescript-eslint/no-require-imports
const webpush = require('web-push') as {
  setVapidDetails: (subject: string, publicKey: string, privateKey: string) => void;
  sendNotification: (
    subscription: { endpoint: string; keys: { p256dh: string; auth: string } },
    payload: string
  ) => Promise<unknown>;
};

export const PUSH_TITLE_APPROVAL_REQUEST = '승인 요청이 도착했습니다';

export interface PushSubscriptionRecord {
  endpoint: string;
  p256dh: string;
  auth: string;
}

export interface SendPushOptions {
  subscriptions: PushSubscriptionRecord[];
  title: string;
  body: string;
  onExpired?: (endpoint: string) => Promise<void>;
  onSuccess?: () => void;
}

/**
 * VAPID 설정을 적용하고 각 구독에 Push 알림을 발송합니다.
 * 발송 실패 시에도 나머지 구독에 계속 발송을 시도합니다.
 * 410 에러(구독 만료) 발생 시 onExpired 콜백을 호출합니다.
 */
export async function sendPushNotifications(options: SendPushOptions): Promise<void> {
  const { subscriptions, title, body, onExpired, onSuccess } = options;

  const VAPID_SUBJECT = process.env.VAPID_SUBJECT ?? '';
  const VAPID_PUBLIC_KEY = process.env.VAPID_PUBLIC_KEY ?? '';
  const VAPID_PRIVATE_KEY = process.env.VAPID_PRIVATE_KEY ?? '';

  if (!VAPID_SUBJECT || !VAPID_PUBLIC_KEY || !VAPID_PRIVATE_KEY) {
    throw new Error(
      '[Push] VAPID 환경변수가 설정되지 않았습니다. VAPID_SUBJECT, VAPID_PUBLIC_KEY, VAPID_PRIVATE_KEY를 확인하세요.'
    );
  }

  webpush.setVapidDetails(VAPID_SUBJECT, VAPID_PUBLIC_KEY, VAPID_PRIVATE_KEY);

  const payload = JSON.stringify({ title, body });

  console.log(`[Push] 발송 시작: ${subscriptions.length}건, title="${title}"`);

  let successCount = 0;
  let failCount = 0;
  let expiredCount = 0;

  for (const subscription of subscriptions) {
    try {
      await webpush.sendNotification(
        {
          endpoint: subscription.endpoint,
          keys: {
            p256dh: subscription.p256dh,
            auth: subscription.auth,
          },
        },
        payload
      );
      successCount++;
      if (onSuccess) {
        onSuccess();
      }
    } catch (error: unknown) {
      if (
        typeof error === 'object' &&
        error !== null &&
        'statusCode' in error &&
        (error as { statusCode: number }).statusCode === 410
      ) {
        expiredCount++;
        console.warn(`[Push] 구독 만료(410): endpoint=${subscription.endpoint}`);
        if (onExpired) {
          await onExpired(subscription.endpoint);
        }
      } else {
        failCount++;
        const statusCode = typeof error === 'object' && error !== null && 'statusCode' in error
          ? (error as { statusCode: number }).statusCode
          : 'unknown';
        console.error(`[Push] 발송 실패: endpoint=${subscription.endpoint}, statusCode=${statusCode}, error=${error}`);
      }
      // 개별 발송 실패는 무시하고 계속 진행
    }
  }

  console.log(`[Push] 발송 완료: 성공=${successCount}, 실패=${failCount}, 만료=${expiredCount}`);
}

File: backend/internal/push/push.go
package push

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"

	webpush "github.com/SherClockHolmes/webpush-go"
	"github.com/dlddu/gatekeeper/backend/internal/models"
)

const TitleApprovalRequest = "승인 요청이 도착했습니다"

type Service struct {
	subject    string
	publicKey  string
	privateKey string
}

func NewService(subject, publicKey, privateKey string) *Service {
	return &Service{subject: subject, publicKey: publicKey, privateKey: privateKey}
}

func (s *Service) configured() error {
	if s.subject == "" || s.publicKey == "" || s.privateKey == "" {
		return fmt.Errorf("[Push] VAPID 환경변수가 설정되지 않았습니다. VAPID_SUBJECT, VAPID_PUBLIC_KEY, VAPID_PRIVATE_KEY를 확인하세요")
	}
	return nil
}

type SendOptions struct {
	Subscriptions []*models.PushSubscription
	Title         string
	Body          string
	// OnExpired is invoked for subscriptions returning HTTP 410.
	OnExpired func(ctx context.Context, endpoint string) error
	// OnSuccess is invoked once per successful delivery.
	OnSuccess func()
}

// Send delivers a notification to each subscription, mirroring the behavior of
// lib/push.ts: failures are logged and skipped, expirations trigger OnExpired.
func (s *Service) Send(ctx context.Context, opts SendOptions) error {
	if err := s.configured(); err != nil {
		return err
	}

	payload, err := json.Marshal(map[string]string{"title": opts.Title, "body": opts.Body})
	if err != nil {
		return fmt.Errorf("marshal push payload: %w", err)
	}

	log.Printf("[Push] 발송 시작: %d건, title=%q", len(opts.Subscriptions), opts.Title)

	successCount, failCount, expiredCount := 0, 0, 0
	for _, sub := range opts.Subscriptions {
		ws := &webpush.Subscription{
			Endpoint: sub.Endpoint,
			Keys: webpush.Keys{
				P256dh: sub.P256dh,
				Auth:   sub.Auth,
			},
		}
		resp, sendErr := webpush.SendNotificationWithContext(ctx, payload, ws, &webpush.Options{
			Subscriber:      s.subject,
			VAPIDPublicKey:  s.publicKey,
			VAPIDPrivateKey: s.privateKey,
			TTL:             30,
		})
		if sendErr != nil {
			failCount++
			log.Printf("[Push] 발송 실패: endpoint=%s, error=%v", sub.Endpoint, sendErr)
			continue
		}
		statusCode := resp.StatusCode
		// Drain and close the response body to allow connection reuse.
		_, _ = io.Copy(io.Discard, resp.Body)
		_ = resp.Body.Close()

		if statusCode == http.StatusGone { // 410: 구독 만료
			expiredCount++
			log.Printf("[Push] 구독 만료(410): endpoint=%s", sub.Endpoint)
			if opts.OnExpired != nil {
				if err := opts.OnExpired(ctx, sub.Endpoint); err != nil {
					log.Printf("[Push] 만료 후 처리 실패: endpoint=%s, error=%v", sub.Endpoint, err)
				}
			}
			continue
		}
		if statusCode >= 200 && statusCode < 300 {
			successCount++
			if opts.OnSuccess != nil {
				opts.OnSuccess()
			}
			continue
		}

		failCount++
		log.Printf("[Push] 발송 실패: endpoint=%s, statusCode=%d", sub.Endpoint, statusCode)
	}

	log.Printf("[Push] 발송 완료: 성공=%d, 실패=%d, 만료=%d", successCount, failCount, expiredCount)
	return nil
}
